cmds/exp/cbmem/io.go

```diff
@@ -1,27 +1,72 @@
-// Copyright 2016-2017 the u-root Authors. All rights reserved
+// Copyright 2016-2021 the u-root Authors. All rights reserved
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
+// +build linux
+
 package main
 
 import (
+	"bytes"
 	"encoding/binary"
+	"fmt"
 	"io"
-	"log"
+	"os"
+	"syscall"
 )
 
+type offsetReader struct {
+	name string
+	base int64
+	r    io.ReaderAt
+}
+
+var _ io.ReaderAt = &offsetReader{}
+
+// ReadAt implements io.ReaderAt.
+// The offset is adjusted by the base. After that
+// it is bytes.Buffer's job to deal with all the
+// corner cases.
+func (o *offsetReader) ReadAt(b []byte, i int64) (int, error) {
```

Comment:
Could this be a io.SectionReader?

Reply:
no it could not. It has to take unadjusted addresses, and that's not what section readers do. Their offsets are 0 relative to the section.